# Analyzing Mobile App Data
1. For this project, we'll pretend we're working as data analysts for a company that builds Android and iOS mobile apps. We make our apps available on Google Play and in the App Store. We only build apps that are free to download and install, and our main source of revenue consists of in-app ads. This means that the number of users of our apps determines our revenue for any given app — the more users who see and engage with the ads, the better.
2. Our goal for this project is to analyze data to help our developers understand what type of apps are likely to attract more users.

## PART 1 - Explore and clean data

## Opening and Exploring the Data

In [1]:
from csv import reader

In [2]:
def extract(data_file):
    opened_file = open(data_file)
    read_file = reader(opened_file)
    app_data = list(read_file)
    return app_data

In [3]:
android_data = extract('/Users/number1/2024/Data/Project1/Android/googleplaystore.csv')
android_header = android_data[0]
android = android_data[1:]
ios_data = extract('/Users/number1/2024/Data/Project1/Appstore/AppleStore.csv')
ios_header = ios_data[0]
ios = ios_data[1:]

In [4]:
def explore_data(data,start=0,end=3,rows_and_columns = False):
    for row in data[start:end]:
            print(row)
            print('\n')
    if rows_and_columns:
        print('number of row:',len(data))
        print('number of column:',len(data[0]))

In [5]:
explore_data(android_data,rows_and_columns= True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


number of row: 10842
number of column: 13


When we look at android_data, there're 10842 rows, 13 column and some useful columns that that fit our aim are: App, Category, Rating, Review, Installs, Type, Genre.

In [6]:
explore_data(ios_data,rows_and_columns= True)

['', 'id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['1', '281656475', 'PAC-MAN Premium', '100788224', 'USD', '3.99', '21292', '26', '4', '4.5', '6.3.5', '4+', 'Games', '38', '5', '10', '1']


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


number of row: 7198
number of column: 17


We have 7197 iOS apps in this data set, and the columns that seem interesting are: 'track_name', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', and 'prime_genre'.

-------------------------------------------------

## Deleting Wrong Data

Recall that at our company, we only build apps that are free to download and install, and we design them for an English-speaking audience. This means that we'll need to do the following:

1. Remove non-English apps like 爱奇艺PPS -《欢乐颂2》电视剧热播.
2. Remove apps that aren't free.

In [7]:
#detect row doesn't have full data
#Android
error1 = 0
count1 = -1
for row in android:
    count1 += 1
    if len(row) != 13:
        error1 += 1
        print('number of columns:',len(row))
        print('\n')
        print('index:', count1)
        print('\n')
        print(android[count1])
        print('\n')
if error1 ==0:
    print('all rows in android have enough columns')


number of columns: 12


index: 10472


['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']




In [8]:
#detect row doesn't have full data
#IOS
error2 = 0
count2 = -1
for row in ios:
    count2 += 1
    if len(row) != 17:
        error2 += 1
        print('number of columns:',len(row))
        print('\n')
        print('index:', count2)
        print('\n')
        print(android[count2])
        print('\n')
if error2 == 0:
    print('all rows in ios have enough columns')


all rows in ios have enough columns


In [9]:
#delete rows that doesn't have enough data
del android[10472]

In [10]:
#check again
error1 = 0
count1 = -1
for row in android:
    count1 += 1
    if len(row) != 13:
        error += 1
        print('number of columns:',len(row))
        print('\n')
        print('index:', count1)
        print('\n')
        print(android[count1])
        print('\n')
if error1 == 0:
    print('all rows have enough columns')

all rows have enough columns


## Removing Duplicate Entries

In [11]:
#Check for duplicate data
def duplicate(data,index):
    unique_apps = []
    duplicate_apps = []
    for row in data:
        a = row[index]
        if a in unique_apps:
            duplicate_apps.append(a)
        else:
            unique_apps.append(a)
    return unique_apps, duplicate_apps


In [12]:
unique_android = duplicate(android,0)[0]
duplicate_android = duplicate(android,0)[1]
print(duplicate_android[:5])
print('\n')
unique_ios = duplicate(ios,0)[0]
duplicate_ios = duplicate(ios,0)[1]
print(duplicate_ios[:5])

['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings']


[]


We see that ios_data doesn't have duplicate data

In [13]:
for row in android:
    name = row[0]
    if name in duplicate_android[:3]:
        print(row)
        print('\n')

['Google My Business', 'BUSINESS', '4.4', '70991', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 24, 2018', '2.19.0.204537701', '4.4 and up']


['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']


['Quick PDF Scanner + OCR FREE', 'BUSINESS', '4.2', '80805', 'Varies with device', '5,000,000+', 'Free', '0', 'Everyone', 'Business', 'February 26, 2018', 'Varies with device', '4.0.3 and up']


['Box', 'BUSINESS', '4.2', '159872', 'Varies with device', '10,000,000+', 'Free', '0', 'Everyone', 'Business', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Google My Business', 'BUSINESS', '4.4', '70991', 'Varies with device', '5,000,000+'

I'll keep row that have the most reviews (because the more reiviews the more recent the data is)

In [14]:
#Find out how many Apps are duplicated
unique_duplicate_apps = []
for name in duplicate_android:
    if name not in unique_duplicate_apps:
        unique_duplicate_apps.append(name)
print(unique_duplicate_apps[:20])
len(unique_duplicate_apps)

['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Zenefits', 'Google Ads', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software', 'MailChimp - Email, Marketing Automation', 'Crew - Free Messaging and Scheduling', 'Asana: organize team projects', 'Google Analytics', 'AdWords Express', 'Accounting App - Zoho Books', 'Invoice & Time Tracking - Zoho']


798

At this step. I'll create a dictionary to lock app's name with max_review of it. And check if it fit my expectation. Lastly, I'll create a new cleaned data which only contains unique app with the most recent review.

In [15]:
highest_review_android = {}
for row in android:
    name = row[0]
    review = float(row[3])
    if name in highest_review_android and review > highest_review_android[name]:
        highest_review_android[name]= review
    elif name not in highest_review_android:
        highest_review_android[name]= review
expected_data = len(android) - len(duplicate_android)       
print('length expected is:', expected_data)
print(len(highest_review_android))


length expected is: 9659
9659


In [16]:
cleaned_android = []
added_android = [] 
for row in android:
    name = row[0]
    review = float(row[3])
    if name not in added_android and review == highest_review_android[name]:
        added_android.append(name) # Vi duplicate data
        cleaned_android.append(row)
print(expected_data)
print(len(cleaned_android))
print(cleaned_android[:5])

9659
9659
[['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'], ['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'], ['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'], ['Pixel Draw - Number Art Coloring Book', 'ART_AND_DESIGN', '4.3', '967', '2.8M', '100,000+', 'Free', '0', 'Everyone', 'Art & Design;Creativity', 'June 20, 2018', '1.1', '4.4 and up'], ['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']]


## Isolating the Free Apps
As we mentioned in the introduction, we only build apps that are free to download and install, and our main source of revenue consists of in-app ads. Our data sets contain both free and non-free apps, and we'll need to isolate only the free apps for our analysis. Below, we isolate the free apps for both our data sets.

In [17]:
#Create a frequence dictionary function.
def freq(data,index):
    freq_dic ={}
    for row in data:
        x = row[index]
        if x in freq_dic:
            freq_dic[x] += 1
        else:
            freq_dic[x] = 1
    return freq_dic

In [18]:
#Create a proportion dictionary function.
def prop(data,index):
    count_dic = freq(data,index)
    proportion_dic ={}
    for x in count_dic:
        proportion = (count_dic[x]/len(data))*100
        proportion_dic[x] = proportion
    return proportion_dic

In [19]:
freq_price_android = freq(cleaned_android,7)
prop_price_android = prop(cleaned_android,7)
print(freq_price_android)
print('\n')
print(prop_price_android)
freq_price_ios = freq(ios,5)
prop_price_ios = prop(ios,5)
print(freq_price_ios)
print('\n')
print(prop_price_ios)

{'0': 8905, '$4.99': 70, '$3.99': 57, '$1.49': 46, '$2.99': 124, '$7.99': 7, '$5.99': 26, '$3.49': 7, '$1.99': 73, '$6.99': 10, '$9.99': 19, '$7.49': 2, '$0.99': 145, '$9.00': 1, '$5.49': 5, '$10.00': 2, '$11.99': 3, '$79.99': 1, '$16.99': 2, '$14.99': 9, '$1.00': 3, '$29.99': 5, '$2.49': 25, '$24.99': 3, '$10.99': 2, '$1.50': 1, '$19.99': 5, '$15.99': 1, '$33.99': 1, '$74.99': 1, '$39.99': 2, '$3.95': 1, '$4.49': 9, '$1.70': 2, '$8.99': 5, '$2.00': 3, '$3.88': 1, '$25.99': 1, '$399.99': 12, '$17.99': 2, '$400.00': 1, '$3.02': 1, '$1.76': 1, '$4.84': 1, '$4.77': 1, '$1.61': 1, '$2.50': 1, '$1.59': 1, '$6.49': 5, '$1.29': 1, '$5.00': 1, '$13.99': 2, '$299.99': 1, '$379.99': 1, '$37.99': 1, '$18.99': 1, '$389.99': 1, '$19.90': 1, '$8.49': 2, '$1.75': 1, '$14.00': 1, '$4.85': 1, '$46.99': 1, '$109.99': 1, '$154.99': 1, '$3.08': 1, '$2.59': 1, '$4.80': 1, '$1.96': 1, '$19.40': 1, '$3.90': 1, '$4.59': 1, '$15.46': 1, '$3.04': 1, '$12.99': 3, '$4.29': 1, '$2.60': 1, '$3.28': 1, '$4.60': 1, '

## Delete Unneeded letter
Because Price include $

so we can not change it to type. Let's fix that by delete the $ letter.

In [20]:
for row in cleaned_android:
    price = row[7]
    price = price.replace('$','')
    row[7] = price

In [21]:
#Remove app that aren't free
def free_app(data,index):
    free_apps = []
    for row in data:
        price = float(row[index])
        if price == 0:
            free_apps.append(row)
    return free_apps


In [22]:
cleaned_free_android = free_app(cleaned_android,7)
explore_data(cleaned_free_android)
print(len(cleaned_free_android))
print('\n')
free_ios = free_app(ios,5)
explore_data(free_ios)
print('\n')
print(len(free_ios))

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


8905


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


['4', '282614216', 'eBay: Best App to Buy, Sell, Save! Online Shopping', '128512000', 'USD', '0', '262

## Removing Non-English Apps
Next step, We'll removing not English App. Remember we use English for the apps we develop at our company, and we'd like to analyze only the apps that are designed for an English-speaking audience. The numbers corresponding to the characters we commonly use in an English text are all in the range 0 to 127, according to the ASCII (American Standard Code for Information Interchange) system. Based on this number range, we can build a function that detects whether a character belongs to the set of common English characters or not. If the number is equal to or less than 127, then the character belongs to the set of common English characters.

If an app name contains a character that is greater than 127, then it probably means that the app has a non-English name.To minimize the impact of data loss, we'll only remove an app if its name has more than three non-ASCII characters

In [23]:
def is_english(string):
    non_english_char = 0
    for char in string:
        if ord(char) >= 127:
            non_english_char += 1
    if non_english_char <= 3:
        return True
    else:
        return False
    

In [24]:
#check
print(is_english('toilathangkhung'))
print(is_english('åååååå´œœ´œ'))

True
False


In [25]:
english_android = []
for row in cleaned_free_android:
    name = row[0]
    if is_english(name):
        english_android.append(row)
explore_data(english_android,0,3,True)
english_ios =[]
print('\n')
for row in free_ios:
    name = row[2]
    if is_english(name):
        english_ios.append(row)
explore_data(english_ios,0,3,True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


number of row: 8864
number of column: 13


['2', '281796108', 'Evernote - stay organized', '158578688', 'USD', '0', '161065', '26', '4', '3.5', '8.2.2', '4+', 'Productivity', '37', '5', '23', '1']


['3', '281940292', 'WeatherBug - Local Weather, Radar, Maps, Alerts', '100524032', 'USD', '0', '188583', '2822', '3.5', '4.5', '5.0.0', '4+', 'Weather', '37', '5', '3', '1']


['4', '282614216', 'eBay: Best App to Buy, Sell, Save! Online Shop

## PART 2 - Analyze
After cleaning the data, we got 8864 apps in google play and 3222 apps in appstore ready to be analyzed.

As we mentioned in the introduction, our aim is to determine the kinds of apps that are likely to attract more users because our revenue is highly influenced by the number of people using our apps.

To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

- Build a minimal Android version of the app, and add it to Google Play.
- If the app has a good response from users, we then develop it further.
- If the app is profitable after six months, we also build an iOS version of the app and add it to the App Store.

Because our end goal is to add the app on both the App Store and Google Play, we need to find app profiles that are successful on both markets. For instance, a profile that might work well for both markets might be a productivity app that makes use of gamification.

Let's begin the analysis by getting a sense of the most common genres for each market. For this, we'll build a frequency table for the prime_genre column of the App Store data set, and the Genres and Category columns of the Google Play data set.

We'll build two functions we can use to analyze the frequency tables:

- One function to generate frequency tables that show percentages
- Another function that we can use to display the percentages in a descending order

In [26]:
final_android = english_android
final_ios = english_ios

## Analyze The Frequency Tables

In [27]:
#We've already create a function to display the percentages prop(data, index)
#Now let's create a function that we can use to display the percentage in a descending order
def display_table(data,index):
    table = prop(data,index)
    display_table = []
    for key in table:
        add =(table[key],key)
        display_table.append(add)
    display_table = sorted(display_table,reverse = True)
    for app in display_table:
        print(app[1],': ',app[0])
    

## Most Common Apps by Genre

We start by examining the frequency table for the prime_genre column of the App Store data set.

In [28]:
#IOS proportion prime_genre
display_table(final_ios,-5)

Games :  58.16263190564867
Entertainment :  7.883302296710118
Photo & Video :  4.9658597144630665
Education :  3.662321539416512
Social Networking :  3.2898820608317814
Shopping :  2.60707635009311
Utilities :  2.5139664804469275
Sports :  2.1415270018621975
Music :  2.0484171322160147
Health & Fitness :  2.0173805090006205
Productivity :  1.7380509000620732
Lifestyle :  1.5828677839851024
News :  1.3345747982619491
Travel :  1.2414649286157666
Finance :  1.1173184357541899
Weather :  0.8690254500310366
Food & Drink :  0.8069522036002483
Reference :  0.5586592178770949
Business :  0.5276225946617008
Book :  0.4345127250155183
Navigation :  0.186219739292365
Medical :  0.186219739292365
Catalogs :  0.12414649286157665


We can see that among the free English apps, more than a half (58.16%) are games. Entertainment apps are close to 8%, followed by photo and video apps, which are close to 5%. Only 3.66% of the apps are designed for education, followed by social networking apps which amount for 3.29% of the apps in our data set.

The general impression is that App Store (at least the part containing free English apps) is dominated by apps that are designed for fun (games, entertainment, photo and video, social networking, sports, music, etc.), while apps with practical purposes (education, shopping, utilities, productivity, lifestyle, etc.) are more rare. However, the fact that fun apps are the most numerous doesn't also imply that they also have the greatest number of users — the demand might not be the same as the offer.

Let's continue by examining the Genres and Category columns of the Google Play data set (two columns which seem to be related).


In [29]:
#Android - Categories proportion
display_table(final_android,1)

FAMILY :  18.907942238267147
GAME :  9.724729241877256
TOOLS :  8.461191335740072
BUSINESS :  4.591606498194946
LIFESTYLE :  3.9034296028880866
PRODUCTIVITY :  3.892148014440433
FINANCE :  3.7003610108303246
MEDICAL :  3.531137184115524
SPORTS :  3.395758122743682
PERSONALIZATION :  3.3167870036101084
COMMUNICATION :  3.2378158844765346
HEALTH_AND_FITNESS :  3.0798736462093865
PHOTOGRAPHY :  2.944494584837545
NEWS_AND_MAGAZINES :  2.7978339350180503
SOCIAL :  2.6624548736462095
TRAVEL_AND_LOCAL :  2.33528880866426
SHOPPING :  2.2450361010830324
BOOKS_AND_REFERENCE :  2.1435018050541514
DATING :  1.861462093862816
VIDEO_PLAYERS :  1.7937725631768955
MAPS_AND_NAVIGATION :  1.3989169675090252
FOOD_AND_DRINK :  1.2409747292418771
EDUCATION :  1.1620036101083033
ENTERTAINMENT :  0.9589350180505415
LIBRARIES_AND_DEMO :  0.9363718411552346
AUTO_AND_VEHICLES :  0.9250902527075812
HOUSE_AND_HOME :  0.8235559566787004
WEATHER :  0.8009927797833934
EVENTS :  0.7107400722021661
PARENTING :  0.6543

The landscape seems significantly different on Google Play: there are not that many apps designed for fun, and it seems that a good number of apps are designed for practical purposes (family, tools, business, lifestyle, productivity, etc.). However, if we investigate this further, we can see that the family category (which accounts for almost 19% of the apps) means mostly games for kids.


![Image](file:///Users/number1/Desktop/Screenshot%202024-02-17%20at%2014.30.09.png)

Even so, practical apps seem to have a better representation on Google Play compared to App Store. This picture is also confirmed by the frequency table we see for the Genres column:


In [30]:
#Android - Genre proportion
display_table(final_android,-4)

Tools :  8.449909747292418
Entertainment :  6.069494584837545
Education :  5.347472924187725
Business :  4.591606498194946
Productivity :  3.892148014440433
Lifestyle :  3.892148014440433
Finance :  3.7003610108303246
Medical :  3.531137184115524
Sports :  3.463447653429603
Personalization :  3.3167870036101084
Communication :  3.2378158844765346
Action :  3.1024368231046933
Health & Fitness :  3.0798736462093865
Photography :  2.944494584837545
News & Magazines :  2.7978339350180503
Social :  2.6624548736462095
Travel & Local :  2.3240072202166067
Shopping :  2.2450361010830324
Books & Reference :  2.1435018050541514
Simulation :  2.0419675090252705
Dating :  1.861462093862816
Arcade :  1.8501805054151623
Video Players & Editors :  1.7712093862815883
Casual :  1.7599277978339352
Maps & Navigation :  1.3989169675090252
Food & Drink :  1.2409747292418771
Puzzle :  1.128158844765343
Racing :  0.9927797833935018
Role Playing :  0.9363718411552346
Libraries & Demo :  0.9363718411552346
Aut

The difference between the Genres and the Category columns is not crystal clear, but one thing we can notice is that the Genres column is much more granular (it has more categories). We're only looking for the bigger picture at the moment, so we'll only work with the Category column moving forward.

Up to this point, we found that the App Store is dominated by apps designed for fun, while Google Play shows a more balanced landscape of both practical and for-fun apps. Now we'd like to get an idea about the kind of apps that have most users.

## Most Popular Apps by Genre on the App Store

One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs and rating for each app genre. For the Google Play data set, we can find this information in the Installs column, but unfortunately for the App Store data set this information is missing. As a workaround, we'll take the total number of user ratings as a proxy, which we can find in the rating_count_tot app.

Below, we create a function to calculate the average number of user ratings per app genre on the App Store:

In [31]:
#Create function for average:
def avg(data,index1,index2):
    table = freq(data,index1)
    display_table = []
    for a in table:
        count = 0
        total = 0
        for b in data:
            value = float(b[index2])
            name = b[index1]
            if name == a:
                count += 1
                total += value
        avg = total/ count
        add = (avg,a)
        display_table.append(add)
    display_table = sorted(display_table,reverse = True)
    for genre in display_table:
        print(genre[1],': ',genre[0])    

In [32]:
#Average count_rating_tot per Genre for IOS
avg(final_ios,-5,6) 

Navigation :  86090.33333333333
Reference :  74942.11111111111
Social Networking :  71548.34905660378
Music :  57326.530303030304
Weather :  52279.892857142855
Book :  39758.5
Food & Drink :  33333.92307692308
Finance :  31467.944444444445
Photo & Video :  28441.54375
Travel :  28243.8
Shopping :  26919.690476190477
Health & Fitness :  23298.015384615384
Sports :  23008.898550724636
Games :  22788.6696905016
News :  21248.023255813954
Productivity :  21028.410714285714
Utilities :  18684.456790123455
Lifestyle :  16485.764705882353
Entertainment :  14029.830708661417
Business :  7491.117647058823
Education :  7003.983050847458
Catalogs :  4004.0
Medical :  612.0


On count_rating_tot average,the top 4 are: Navigation, Reference, Social Networking and Music but when we look it closely, this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together:

In [33]:
#Create funcing to check information in desc order
def check(data,name,index1 =-5,index2 =6,index3 =2,start =0,end =20):
    display_table = []
    for row in data:
        value1 = row[index1]
        value2 = float(row[index2])
        value3 = row[index3]
        if value1 == name:
            add =(value2,value3)
            display_table.append(add)
    display_table = sorted(display_table,reverse = True)
    for app in display_table[start:end]:
        print(app[1],': ',app[0])

In [34]:
#check for Navigation
check(final_ios,'Navigation')

Waze - GPS Navigation, Maps & Real-time Traffic :  345046.0
Google Maps - Navigation & Transit :  154911.0
Geocaching® :  12811.0
CoPilot GPS – Car Navigation & Offline Maps :  3582.0
ImmobilienScout24: Real Estate Search in Germany :  187.0
Railway Route Search :  5.0


Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating:

In [35]:
#check for Reference
check(final_ios,'Reference')

Bible :  985920.0
Dictionary.com Dictionary & Thesaurus :  200047.0
Dictionary.com Dictionary & Thesaurus for iPad :  54175.0
Google Translate :  26786.0
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran :  18418.0
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition :  17588.0
Merriam-Webster Dictionary :  16849.0
Night Sky :  12122.0
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) :  8535.0
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools :  4693.0
GUNS MODS for Minecraft PC Edition - Mods Tools :  1497.0
Guides for Pokémon GO - Pokemon GO News and Cheats :  826.0
WWDC :  762.0
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free :  718.0
VPN Express :  14.0
Real Bike Traffic Rider Virtual Reality Glasses :  8.0
教えて!goo :  0.0
Jishokun-Japanese English Dictionary & Translator :  0.0


The same pattern applies to social networking apps, where the average number is heavily influenced by a few giants like Facebook, Pinterest, Skype, etc.

In [36]:
#Check for Social Networking
check(final_ios,'Social Networking')

Facebook :  2974676.0
Pinterest :  1061624.0
Skype for iPhone :  373519.0
Messenger :  351466.0
Tumblr :  334293.0
WhatsApp Messenger :  287589.0
Kik :  260965.0
ooVoo – Free Video Call, Text and Voice :  177501.0
TextNow - Unlimited Text + Calls :  164963.0
Viber Messenger – Text & Call :  164249.0
Followers - Social Analytics For Instagram :  112778.0
MeetMe - Chat and Meet New People :  97072.0
We Heart It - Fashion, wallpapers, quotes, tattoos :  90414.0
InsTrack for Instagram - Analytics Plus More :  85535.0
Tango - Free Video Call, Voice and Chat :  75412.0
LinkedIn :  71856.0
Match™ - #1 Dating App. :  60659.0
Skype for iPad :  60163.0
POF - Best Dating App for Conversations :  52642.0
Timehop :  49510.0


Same applies to music apps, where a few big players like Pandora, Spotify, and Shazam heavily influence the average number:

In [37]:
#Check for Music:
check(final_ios,'Music')

Pandora - Music & Radio :  1126879.0
Spotify Music :  878563.0
Shazam - Discover music, artists, videos & lyrics :  402925.0
iHeartRadio – Free Music & Radio Stations :  293228.0
SoundCloud - Music & Audio :  135744.0
Magic Piano by Smule :  131695.0
Smule Sing! :  119316.0
TuneIn Radio - MLB NBA Audiobooks Podcasts Music :  110420.0
Amazon Music :  106235.0
SoundHound Song Search & Music Player :  82602.0
Sonos Controller :  48905.0
Bandsintown Concerts :  30845.0
Karaoke - Sing Karaoke, Unlimited Songs! :  28606.0
My Mixtapez Music :  26286.0
Sing Karaoke Songs Unlimited with StarMaker :  26227.0
Ringtones for iPhone & Ringtone Maker :  25403.0
Musi - Unlimited Music For YouTube :  25193.0
AutoRap by Smule :  18202.0
Spinrilla - Mixtapes For Free :  15053.0
Napster - Top Music & Radio :  14268.0


Next, We'll looking into average Rating to find more information

In [38]:
#average rating per Genre for IOS
avg(final_ios,-5,8)

Catalogs :  4.125
Games :  4.037086446104589
Productivity :  4.0
Business :  3.9705882352941178
Shopping :  3.9702380952380953
Music :  3.946969696969697
Photo & Video :  3.903125
Navigation :  3.8333333333333335
Health & Fitness :  3.769230769230769
Reference :  3.6666666666666665
Education :  3.635593220338983
Food & Drink :  3.6346153846153846
Social Networking :  3.5943396226415096
Entertainment :  3.5393700787401574
Utilities :  3.5308641975308643
Travel :  3.4875
Weather :  3.482142857142857
Lifestyle :  3.411764705882353
Finance :  3.375
News :  3.244186046511628
Book :  3.0714285714285716
Sports :  3.0652173913043477
Medical :  3.0


On average rating, Top 4 are: Catalogs, Games, Productivity, Business. Let's look back on the previous average count_rating_tot we notice that Game have the most couting rating than the other three.
As we look at average rating and average count rating. The top 4 count_rating always have 1 or 2 giant that dominate the market which we've mentioned above and our budget doesn't give us an ability to compete with them Because of that Games same to be our best options. Let's look at it :

In [39]:
#check for Game
check(final_ios,'Games')

Clash of Clans :  2130805.0
Temple Run :  1724546.0
Candy Crush Saga :  961794.0
Angry Birds :  824451.0
Subway Surfers :  706110.0
Solitaire :  679055.0
CSR Racing :  677247.0
Crossy Road - Endless Arcade Hopper :  669079.0
Injustice: Gods Among Us :  612532.0
Hay Day :  567344.0
PAC-MAN :  508808.0
DragonVale :  503230.0
Head Soccer :  481564.0
Despicable Me: Minion Rush :  464312.0
The Sims™ FreePlay :  446880.0
Sonic Dash :  418033.0
8 Ball Pool™ :  416736.0
Tiny Tower - Free City Building :  414803.0
Jetpack Joyride :  405647.0
Bike Race - Top Motorcycle Racing Games :  405007.0


Leading the ranking board is Clash of Clans - 2M1 which follwed by Temple Run - 1m7, Candy Rush Saga- 960k. As we look at Top-20 board, we notice that most of these game is either from endless run game (Temple Run, Subway Surfers,...) or puzzle game (Candy Crush Saga, Angry Birds,...) and the same are true in Google play:

For the Google Play market, we actually have data about the number of installs, so we should be able to get a clearer picture about genre popularity. However, the install numbers don't seem precise enough — we can see that most values are open-ended (100+, 1,000+, 5,000+, etc.):

First we need to clean the data. Because it's not ready to be use

In [40]:
print(freq(final_android,5))

{'10,000+': 904, '5,000,000+': 605, '50,000,000+': 204, '100,000+': 1024, '50,000+': 423, '1,000,000+': 1394, '10,000,000+': 935, '5,000+': 400, '500,000+': 493, '1,000,000,000+': 20, '100,000,000+': 189, '1,000+': 744, '500,000,000+': 24, '500+': 288, '100+': 613, '50+': 170, '10+': 314, '1+': 45, '5+': 70, '0+': 4, '0': 1}


One problem with this data is that is not precise. For instance, we don't know whether an app with 100,000+ installs has 100,000 installs, 200,000, or 350,000. However, we don't need very precise data for our purposes — we only want to get an idea which app genres attract the most users, and we don't need perfect precision with respect to the number of users.

We're going to leave the numbers as they are, which means that we'll consider that an app with 100,000+ installs has 100,000 installs, and an app with 1,000,000+ installs has 1,000,000 installs, and so on.

To perform computations, however, we'll need to convert each install number to float — this means that we need to remove the commas and the plus characters, otherwise the conversion will fail and raise an error. We'll do this directly in the loop below, where we also compute the average number of installs for each genre (category)

In [41]:
for row in final_android:
    install = row[5]
    install = install.replace('+','')
    install = install.replace(',','')
    row[5] = install
print(freq(final_android,5))

{'10000': 904, '5000000': 605, '50000000': 204, '100000': 1024, '50000': 423, '1000000': 1394, '10000000': 935, '5000': 400, '500000': 493, '1000000000': 20, '100000000': 189, '1000': 744, '500000000': 24, '500': 288, '100': 613, '50': 170, '10': 314, '1': 45, '5': 70, '0': 5}


In [42]:
check(final_android,'GAME',1,5,0)

Subway Surfers :  1000000000.0
Temple Run 2 :  500000000.0
Pou :  500000000.0
My Talking Tom :  500000000.0
Candy Crush Saga :  500000000.0
slither.io :  100000000.0
Zombie Tsunami :  100000000.0
Yes day :  100000000.0
Vector :  100000000.0
Trivia Crack :  100000000.0
Traffic Racer :  100000000.0
Temple Run :  100000000.0
Talking Tom Gold Run :  100000000.0
Super Mario Run :  100000000.0
Sonic Dash :  100000000.0
Sniper 3D Gun Shooter: Free Shooting Games - FPS :  100000000.0
Smash Hit :  100000000.0
Skater Boy :  100000000.0
Shadow Fight 2 :  100000000.0
Score! Hero :  100000000.0


Even though these games are top 20 in number of installation, doesn't mean it have a high rating. Let's check that:

In [43]:
#check rating IOS
i_list = ['Temple Run','Angry Birds','Subway Surfers','Candy Crush Saga']
for row in final_ios:
    app_name = row[2]
    user_rating = row[8]
    if app_name in i_list:
        print(app_name,'-',user_rating)

Angry Birds - 4.5
Temple Run - 4.5
Subway Surfers - 4.5
Candy Crush Saga - 4.5


In [44]:
#Check rating Android:
a_list = ['Subway Surfers','Temple Run 2','Temple Run','Candy Crush Saga','Zombie Tsunami','Angry Birds']
for row in final_android:
    app_name = row[0]
    rating = row[2]
    if app_name in a_list:
        print(app_name,'-',rating)

Candy Crush Saga - 4.4
Subway Surfers - 4.5
Zombie Tsunami - 4.4
Temple Run 2 - 4.3
Temple Run - 4.3


Look like these apps are not popular but they're also adored on both Ios and Android.
However, this niche seems to be dominated by top 5 apps, so it's probably not a good idea to build similar apps since there'll be some significant competition. But there're seem so be some potential because the current game is still simple. So it's a great idea to develope it.

## Conclusion

In this project, we analyzed data about the App Store and Google Play mobile apps with the goal of recommending an app profile that can be profitable for both markets.
We concluded that we could build either endless run game like (Surway Surfers, Temple Run,...) or a puzzel game like (Candy Crush Sage,...) which can be profitable for both the Google Play and the App Store markets. The markets are already full of these games, so we need to add some special features besides the raw version of the games This might include: adding some impressive fashion items and some little minigame in it, ranking reward for top score players.